aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -239,6 +240,159 @@ Tensor wrapped_scalar_tensor(
+    .enforce_safe_casting_to_output(false)
+    .check_all_same_dtype(false)
+    .build();
+  AT_DISPATCH_INTEGRAL_TYPES(elements.scalar_type(), "isin_default", [&]() {
+    iter.for_each([&](char** data, const int64_t* strides, int64_t n) {
+      auto* out_bytes = data[0];
+      const auto* element_bytes = data[1];
+      const auto* test_element_bytes = test_elements_flat.data_ptr();
+
+      for (auto i = 0; i < n; ++i) {
+        auto* out_data = reinterpret_cast<bool*>(out_bytes);
+        const auto element_val = *reinterpret_cast<const scalar_t*>(element_bytes);
+
+        for (auto j = 0; j < test_elements_flat.numel(); ++j) {
+          if (element_val == *(static_cast<const scalar_t*>(test_element_bytes) + j)) {
+            *out_data = !invert;
+            break;
+          }
+        }
+
+        out_bytes += strides[0];
+        element_bytes += strides[1];
+      }
+    });
+  });
+}
+
+// Sorting-based algorithm for isin(); used when the number of test elements is large.
+// 1. Concatenate unique elements with unique test elements in 1D form. If
+//    assume_unique is true, skip calls to unique().
+// 2. Stable sort all elements, maintaining order indices to reverse the
+//    operation. Stable sort is necessary to keep elements before test
+//    elements within the sorted list.
+// 3. Create a mask for locations of adjacent duplicate values within the
+//    sorted list. Duplicate values are in both elements and test elements.
+// 4. Reorder the mask to match the pre-sorted element order.
+// 5. Index the mask to match the pre-unique element order. If
+//    assume_unique is true, just take the first N items of the mask,
+//    where N is the original number of elements.
+static void isin_sorting(const Tensor& elements, const Tensor& test_elements,
+    bool assume_unique, bool invert, Tensor& out) {
+  Tensor elements_flat, test_elements_flat, unique_order;
+  if (assume_unique) {
+    elements_flat = elements.ravel();
+    test_elements_flat = test_elements.ravel();
+  } else {
+    std::tie (elements_flat, unique_order) = at::_unique(
+        elements, /*sorted=*/ false, /*return_inverse=*/ true);
+    std::tie (test_elements_flat, std::ignore) = at::_unique(test_elements, /*sorted=*/ false);
+  }
+
+  Tensor all_elements = at::_cat({elements_flat, test_elements_flat});
+  Tensor sorted_elements, sorted_order;
+  std::tie (sorted_elements, sorted_order) = all_elements.sort(
+      /*stable=*/ true, /*dim=*/ 0, /*descending=*/ false);
+
+  Tensor duplicate_mask = at::empty(all_elements.sizes(),
+    TensorOptions(elements.device()).dtype(ScalarType::Bool));
+  auto iter = TensorIteratorConfig()
+    .set_check_mem_overlap(true)
+    .add_output(duplicate_mask)
+    .add_input(sorted_elements)
+    .cast_common_dtype_to_outputs(false)
+    .enforce_safe_casting_to_output(false)
+    .check_all_same_dtype(false)
+    .build();
+  AT_DISPATCH_INTEGRAL_TYPES(elements.scalar_type(), "isin_sorting", [&]() {
```

Comment:
Same question here: which composite ops are we missing?

Reply:
@mruberry This is actually `in1d` algorithm, I might have a wrong impression that we are not going to support in1d directly, so we just name this `isin`. Should we support `in1d`? and make this `isin` as a composite function?

Reply:
Outside of the `in1d()` vs. `isin()` concern, I'm pretty sure it's possible to implement this part without `TensorIterator`. [Numpy does it](https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/lib/arraysetops.py#L598-L601) using `==` / `!=` on slices and concatenating the result with a final value.

I found the `TensorIterator`-based approach to be straightforward (just check adjacent value equality) and it avoids the extra tensor instantiations from slicing / concatenation.

Reply:
I can believe that implementing a new kernel might be faster, and implementing a fast path is neat, but should we prefer CUDA support over a fast path? We can also have our cake and eat it, too, by implementing the op as a composite and then having a fast path on the CPU only, for example, that uses a new kernel.

Reply:
By fast path, do you mean the sorting-based algorithm?